### ISSUE TITLE
Provide an API to exclude or overload AppThemeProvider

### ISSUE THREAD

**Issue by taras** [backstage/backstage#7103]
## Context

On the HP DevEx project, we're using the default Backstage frontend app for the admin interface. The frontend that users see is a different React app in the same monorepo. Each major section of the UI is a separate package that uses the Backstage plugin template. We use `@backstage/cli plugin:serve` to run each section separately while doing development. We use the `dev.tsx` file to configure the development environment for each section. Here is what the bootstrap mechanism looks like,

```tsx
import React from 'react';
import ReactDOM from 'react-dom';
import { createApp } from '@backstage/core-app-api';
import { Section } from '../src';

const app = createApp();
const BackstageProvider = app.getProvider();

const HpDevExApp = ({ children }) => {
   const wrap  = () => {
     // our setup goes here which uses config from app-config.yaml
   }; 
   return wrap(children);
}

const DevApp: FC<{
  children: ReactNode;
}> = ({ children }: { children: ReactNode }) => {
  return (
    <Suspense fallback={<p>Loading...</p>}>
      <BackstageProvider>
        <HpDevExApp>{children}</HpDevExApp>
      </BackstageProvider>
    </Suspense>
  );
};


ReactDOM.render(
  <DevApp>
    <Section />
  </DevApp>,
  document.getElementById('root'),
);
```

The benefit of this setup is that we're able to use all of the Backstage APIs but use HP's internal component framework. The problem is that [`<CssBaseline />`](https://github.com/backstage/backstage/blob/4045eaab952fec349e443990330bd1ac5b2c9c18/packages/core-app-api/src/app/AppThemeProvider.tsx#L94) messes with CSS from HP's component framework. I would really like to eliminate `AppThemeProvider` or replace it with another provider.

Currently, we created a patch to remove CSS provided by `<CssBaseline />` but it messes with the CSS of our Backstage admin site because that site actually needs those styles.

## Feature Suggestion

Make it possible to replace or remove `AppThemeProvider` from the context tree. 

## Possible Implementation

The possible API could be something like this,

```tsx
const app = createApp();
const BackstageProvider = app.getProvider({ AppThemeProvider: ({ children }) =>  children });
```

We can contribute the change if Backstage team can suggest an API.

**Comment by Rugvip** [comment]
The way to customize the app theme is via [`createApp`](https://backstage.io/docs/getting-started/app-custom-theme#using-your-custom-theme), similar to how the config works.

When it comes to getting rid of `CssBaseline` I'm really not sure if that's a good idea 😅. We could however enable it by making the `AppThemeProvider` a customizable app component, I don't think that would be completely insane given that it also houses the logic for the default theme selection logic.

**Comment by taras** [comment]
@Rugvip would you accept a PR that introduces the ability to customize the theme as I described above? If not, what API would you accept? We can put together a PR based on your response.

**Comment by Rugvip** [comment]
@taras thank you!

I'd want to go with the app component route, both because that's where we collect these kind of overrides, and also because I hope to get rid of the `getProvider` API eventually.

So an addition here: https://github.com/backstage/backstage/blob/c9ce2ba317b6860b06a241dc9ece6ad2e5f73dca/packages/core-app-api/src/app/types.ts#L69

Ends up looking something like this:

```ts
const app = createApp({
  components: {
    ThemeProvider: ...
  }
})
```

Needs to be optional ofc, and note the dropped `App*` prefix as I think it's redundant there.

**Comment by taras** [comment]
@Rugvip perfect! We'll work with this.

@minkimcello let's work on this together when you get a chance.

**Comment by taras** [comment]
Closed in #7201
